### ISSUE TITLE
config: partial rejections in incremental xDS

### ISSUE THREAD

**Issue by fredlas** [envoyproxy/envoy#5739]
In the review of #5466, the issue of partial rejections (when some resources have errors, but you still want to apply the rest and tell the xDS server you did so) came up. It was previously requested/discussed in #3079, but the current thinking is that *only* incremental xDS will have this feature, so I figured it deserved a new issue.

The incremental xDS gRPC protocol, as defined by [IncrementalDiscoveryRequest](https://github.com/envoyproxy/envoy/blob/master/api/envoy/api/v2/discovery.proto#L124), does not currently support partial rejections; there is just a single "here is why everything failed" grpc::Status field, just as in standard xDS. How should this proto support the partial rejection feature?

@htuch and I briefly discussed how this might be done, and saw two options so far:

- Replace the grpc::Status error_detail field with a repeated grpc::Status field, expected to be as long as the number of resources in the previous IncrementalDiscoveryResponse. Entries correspond by index to the resources in that previous message.
- Have a map from resource names to errors. This approach would be complicated by the fact that resource name lives inside an opaque Any.

There is also the question of whether we want to be able to distinguish between "each and every resource individually had errors" and "your whole update attempt had an error", i.e. whether the existing error_detail field should remain alongside the new partial rejection mechanism.

**Comment by htuch** [comment]
@costinm please upvote with Istio context as discussed today.

**Comment by costinm** [comment]
This is one of the biggest problems experienced by users with TLS certificates - we have no good way to 
know when a cert has been distributed to all workloads, so when the LDS config is pushed to envoy the entire update is rejected. Worse - if envoy restarts, it will fail to get an initial config ( because one of the 
certs is missing ). We also saw problems on mesh expansion, where one of the port was in use by an app - rejecting the entire config. Same problem happens for CDS with the certificates missing. 

I think this is by far the most useful fix I would like to see in envoy based on issues our users
experienced.  

In addition, the actual cause of the NACK is hard to parse automatically - at the very least it should
include the names of listeners/clusters that are broken in an easy to process format.

**Comment by htuch** [comment]
@costinm thanks for sharing the Istio perspective. https://github.com/envoyproxy/envoy/pull/5806 just merged, which might provide some improvement in insight into which resources are failing.

**Comment by htuch** [comment]
FWIW, looking at the new incremental implementation in `CdsApiImpl` , we're already doing a partial accept, so this issue is really just tracking making it a first class concern on the wire, i.e. being able to provide feedback at the per-resource granularity rather than a comma-separated string.

**Comment by costinm** [comment]
Thanks for the update - this is great on CDS side, but we'll also need something on LDS. Ingress needs to be configured with certs for multiple domains - and as I mentioned it's tricky for the XDS server to know if an envoy has the certs deployed. 
 
We are switching to SDS for secrets - I'm not sure if LDS failing to load a cert via SDS will result in the full LDS rejection, but we have a report that if the SDS UDS socket is missing we do have problems.

**Comment by htuch** [comment]
@fredlas what are your thoughts on how we should proceed with this one?

**Comment by fredlas** [comment]
As you mentioned early, the implementation details of adding per-resource rejection to the protocol should be almost trivial.

There is the issue of what should actually be done. I remember there was the pretty detailed offline discussion among a few people here, where the issue of resource dependencies was raised - basically, if you need ADS to enforce a particular ordering of your updates, then you need those same dependencies reflected when handling failures. (I think that's a reasonable summary?)

I am vaguely imagining a relatively clean bit of work (in particular with no xDS protocol changes (although maybe some changes to the resource protos themselves)) to accomplish that. In the "XdsGrpcContext" implementation I'm working on, you would also maintain some mapping that tells you "when you process a type_url1 update failure for resource X , reach into the type_url2 subscription to mark Y as unavailable, as well as type_url3 Z".

The good news is that it should fit very cleanly on top of what I'm currently building, rather than needing to be built in from the start. So we can keep that design in mind, while holding off on actually doing it until the current work is done, without hassle.

**Comment by htuch** [comment]
@fredlas how are resource dependencies different when we have partial rejection? In existing ADS, you can have messages [X, Y] queued up, reject X and you still process Y. Now all we're saying is we do a partial reject of X and still process Y. I do remember some discussion around this, but not the particulars.

**Comment by fredlas** [comment]
Oh, I see what you're saying, and you're right, this issue is orthogonal to partial rejections. You could have all-or-nothing rejections and still have this problem. In that case that problem should get its own issue.

**Comment by htuch** [comment]
@fredlas yep, sounds good to open a separate issue, please tag me.

**Comment by htuch** [comment]
@fredlas I've tagged this help wanted. We probably want to nail this down in the context of the v3 API bump for the discovery services and incremental support (I've listed it in https://github.com/envoyproxy/envoy/issues/6271). If you have bandwidth, you're welcome to it, otherwise folks working on API topics should pick this up.

**Comment by htuch** [comment]
CC @envoyproxy/api-shepherds @envoyproxy/udpa-wg

**Comment by fredlas** [comment]
Yeah, probably I shouldn't get involved; I really need to be getting on to my new project!

I think this should be pretty simple: the current code is already concatenating errors per resource into a single string, so instead just feed them into the map option mentioned in the original post. Need to decide what to do with the existing singular error field (maybe it could remain the concatenated string? or maybe change its meaning to "something went wrong more generally than just 'one or more resources were bad'"?)

**Comment by markdroth** [comment]
Is this still needed if we implement #11396?  In general, it seems like the mechanism described there might essentially replace the current ACK/NACK mechanism completely in terms of determining the overall health of the client's config.  However, depending on how we do it, it might not address the use-case of needing to know when a client has actually switched to a new, valid config.

**Comment by htuch** [comment]
@markdroth I see #11396 as providing a way to reflect asynchronously changes to config state. The partial rejection is for synchronous rejections and provides immediate management server / config pipeline feedback.

**Comment by abeyad** [comment]
> Unfortunately, this leaves the control plane without a way to know which specific resources were accepted by the client. It would definitely be good to add a mechanism to the transport protocol to be able to NACK individual resources instead of responses as a whole.

Why would this matter to the management server for SotW, given that it is required to send all resources for the given version in each DiscoveryResponse anyway?

**Comment by markdroth** [comment]
> Why would this matter to the management server for SotW, given that it is required to send all resources for the given version in each DiscoveryResponse anyway?

First, that's true only for LDS and CDS, as per https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#grouping-resources-into-responses.

Second, even if the management server always sends every resource in every response, I don't see how that helps.  If the management server sends a response containing resources A and B, and then the client NACKs that response, the management server has no way of knowing if the client is rejecting only A, only B, or both.

**Comment by markdroth** [comment]
> My question was more along the lines of how does it effect the management server's behavior if it knows which were rejected

I don't think this info will generally change the management server's behavior with regard to the client.  However, it's very useful for the management server to track this information for monitoring or debugging purposes.  As just one example, the management server can expose this info via CSDS.

I don't think this actually has anything to do with SotW vs. incremental or with how resources are grouped into responses.  It's the same issue regardless of those variables.

> If it gets a NACK for the full response, which is the current behavior, then the management server is forced to also send A at version 1 too, even though the client is already happily using it.

No, I don't think that's right.  The question of whether the client ACKs or NACKs a resource does not affect when the server sends it at all -- and that's true regardless of whether the NACK is per-response or per-resource.

The server's job is to always send the most recent version of the resource that it has to the client, regardless of whether the client accepts the resource.  If the client rejects the resource, resending the same resource is not going to accomplish anything, because the client will just reject it again.

Today, in the incremental protocol, the server does not need to resend a resource unless that resource changes -- and that's true regardless of whether the client rejected it.  Changing from per-response to per-resource NACKs will not affect that behavior.

Today, in the SotW protocol, for LDS and CDS, the server *does* need to resend the resource if some other resource changes -- and that's true regardless of whether the client rejected it.  Changing from per-response to per-resource NACKs (which we're not actually planning to do for SotW) would not affect that behavior.
